### ISSUE TITLE
Verify compliant metric SDK specification implementation: MeterProvider/Shutdown

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go#3643]
- [x] Identify all the normative requirements, recommendations, and options [the specification] defines as comments to this issue
- [x] Ensure the [current metric SDK implementation] is compliant with these normative requirements, recommendations, and options in those comments.

[the specification]: https://github.com/open-telemetry/opentelemetry-specification/blob/0c5bfa999a5310801d5c36d2a82cbcd0e32ca9cb/specification/metrics/sdk.md#shutdown
[current metric SDK implementation]: https://github.com/open-telemetry/opentelemetry-go/tree/v1.12.0/sdk/metric

**Comment by MrAlias** [comment]
> `Shutdown` MUST be called only once for each `MeterProvider` instance.

This is a normative statement for the user of a `MeterProvider` it is not something we are in the scope of compliance for.

Our implementation does handle the case where a user calls shutdown multiple times by using a `sync.Once`.

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/config.go#L77

**Comment by MrAlias** [comment]
> `Shutdown` SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.

An error is returned from the method:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/provider.go#L113

**Comment by MrAlias** [comment]
> `Shutdown` SHOULD complete or abort within some timeout.

The method is documented to comply with this recommendation:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/provider.go#L107-L110

The behavior though is going to depend on the Readers and exporters as this context is passed on to them and the cancel logic is expected to be applied there.

The manual reader seems to comply as it doesn't have any blocking operations (it has nowhere to flush a final collection so it just shuts down):

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/manual_reader.go#L97-L113

The periodic reader does a final collection and then shuts down the exporter. Each of these operations pass the context:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/periodic_reader.go#L322-L324

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/periodic_reader.go#L329

The exporter behavior is out of scope for the `MeterProvider` specification. It will be evaluated it its own issue (https://github.com/open-telemetry/opentelemetry-go/issues/3666).

The collection is going to call the pipeline produce. That handles the context timeouts according to the specification:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/pipeline.go#L125C30-L153

It will also call the external metric producers (bridges):

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/periodic_reader.go#L268

These are out of scope of the `MeterProvider` specification.

**Comment by MrAlias** [comment]
> `Shutdown` MAY be implemented as a blocking API or an asynchronous API which notifies the caller via a callback or an event.

`Shutdown` is a synchronous method call:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/provider.go#L97-L118

**Comment by MrAlias** [comment]
> [OpenTelemetry SDK](../overview.md#sdk) authors MAY decide if they want to make the shutdown timeout configurable.

Timeouts are handled via context timeouts.

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/provider.go#L107-L110

**Comment by MrAlias** [comment]
> `Shutdown` MUST be implemented at least by invoking `Shutdown` on all registered [MetricReader](#metricreader) and [MetricExporter](#metricexporter) instances.

Indeed it is:

https://github.com/open-telemetry/opentelemetry-go/blob/b2246d58650bc6db04459353843250ee3bba77fc/sdk/metric/config.go#L25-L83

**Comment by MrAlias** [comment]
Outstanding issues:

- [ ] https://github.com/open-telemetry/opentelemetry-go/issues/4148

**Comment by pellared** [comment]
@MrAlias I think this issue can be resolved.
